crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs

```diff
@@ -0,0 +1,604 @@
+            };
+
+            let Some(inflight) = inflight else {
+                // connection was dropped.
+                break AppendError::retryable(NetworkError::ConnectionClosed);
+            };
+
+            let RemoteInflightAppend {
+                rpc_token,
+                commit_resolver,
+                permit: _permit,
+            } = inflight;
+
+            let appended = tokio::select! {
+                incoming = rpc_token.recv() => {
+                    incoming.map_err(AppendError::Shutdown)
+                },
+                _ = &mut closed => {
+                    Err(AppendError::retryable(NetworkError::ConnectionClosed))
+                }
+            };
+
+            let appended = match appended {
+                Ok(appended) => appended.into_body(),
+                Err(err) => {
+                    // this can only be a terminal error (either shutdown or connection is closing)
+                    commit_resolver.error(err.clone());
+                    break err;
+                }
+            };
+
+            // good chance to update known global tail
+            if let Some(offset) = appended.known_global_tail {
+                known_global_tail.notify_offset_update(offset);
+            }
+
+            // handle status of the response.
+            match appended.status {
+                SequencerStatus::Ok => {
+                    commit_resolver.offset(appended.first_offset);
+                }
+                SequencerStatus::Malformed => {
+                    // While the malformed status is non-terminal for the connection
+                    // (since only one request is malformed),
+                    // the AppendError for the caller is terminal
+                    commit_resolver.error(AppendError::terminal(Malformed));
+                }
+                SequencerStatus::Sealed => {
+                    // A sealed status returns a terminal error since we can immediately cancel
+                    // all inflight append jobs.
+                    commit_resolver.sealed();
+                    break AppendError::Sealed;
+                }
+            }
+        };
+
+        // close channel to stop any further appends calls on the same connection
+        rx.close();
+
+        // Drain and resolve ALL pending appends on this connection.
+        //
+        // todo(azmy): The order of the RemoteInflightAppend's on the channel
+        // does not necessary matches the actual append calls. This is
+        // since sending on the connection and pushing on the rx channel is not an atomic
+        // operation. Which means that, it's possible when we are draining
+        // the pending requests here that we also end up cancelling some inflight appends
+        // that has already received a positive response from the sequencer.
+        //
+        // For now this should not be a problem since they can (possibly) retry
+        // to do the write again later.
```

Comment:
Which means that we effectively produce in at least once manner to Bifrost, right?

Reply:
@tillrohrmann That is correct. Application layer is responsible for running deup. Bifrost guarantees that committed records have accomplished a write quorum. But failed writes can still show up in the logs. For example during seal. Hence application layer must take care of duplicates